sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala

```diff
@@ -633,6 +645,17 @@ private[hive] class HiveClientImpl(
         // whose specs are supersets of this partial spec. E.g. If a table has partitions
         // (b='1', c='1') and (b='1', c='2'), a partial spec of (b='1') will match both.
         val parts = client.getPartitions(hiveTable, s.asJava).asScala
+        // Check whether the partition we are going to drop is empty.
+        // We make a dummy one for the empty partition. See [SPARK-29786] for more details.
```

Comment:
is this how hive resolve the problem?

Reply:
> is this how hive resolve the problem?

Yes, It's the same method as Hive uses.

Reply:
Isn't it bad for performance? i.e. you call `fs.exists` and `fs.listStatus` for each partition.

Reply:
> Isn't it bad for performance? i.e. you call `fs.exists` and `fs.listStatus` for each partition.

Yes, but only affect `drop partitions`. And it won't take much time to do the check while dropping.

Reply:
Can you point to the Hive source code that does the same thing? i.e. create a dummy directory before dropping the partition.

Reply:
> Can you point to the Hive source code that does the same thing? i.e. create a dummy directory before dropping the partition.

In Hive 1.x, it's like [this](https://github.com/apache/hive/blob/6002c510113d9a6aa87159c7386f2a8a4747405b/ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java#L3415-L3420).

Reply:
is it for DROP PARTITION?

Reply:
> is it for DROP PARTITION?

No, it will check every query before executing. Maybe it's better to do the check before all queries?

Reply:
Does Spark have a problem to do table scan when partition directory not exist?

Reply:
> Does Spark have a problem to do table scan when partition directory not exist?

Yes, it will throw `java.io.FileNotFoundException`.